src/library/compiler/llvm.cpp

```diff
@@ -0,0 +1,1242 @@
+}
+
+// == LLVM <-> Lean: ContextRef ==
+
+static lean_external_class *g_Context_class = nullptr;
+
+static inline lean_object *Context_to_lean(LLVMContextRef s) {
+  if (g_Context_class == nullptr) {
+    g_Context_class =
+        lean_register_external_class(donothing_finalize, donothing_foreach);
+  }
+  return lean_alloc_external(g_Context_class, s);
+}
+
+static inline LLVMContextRef lean_to_Context(b_lean_obj_arg s) {
+  return static_cast<LLVMContextRef>(lean_get_external_data(s));
+}
+
+// == LLVM <-> Lean: TypeRef ==
+
+static lean_external_class *g_Type_class = nullptr;
+
+static inline lean_object *Type_to_lean(LLVMTypeRef s) {
+  if (g_Type_class == nullptr) {
+    g_Type_class =
+        lean_register_external_class(donothing_finalize, donothing_foreach);
+  }
+  return lean_alloc_external(g_Type_class, s);
+}
+
+static inline LLVMTypeRef lean_to_Type(b_lean_obj_arg s) {
+  return static_cast<LLVMTypeRef>(lean_get_external_data(s));
+}
+
+// == LLVM <-> Lean: ValueRef ==
+
+static lean_external_class *g_Value_class = nullptr;
+
+static inline lean_object *Value_to_lean(LLVMValueRef s) {
+  if (g_Value_class == nullptr) {
+    g_Value_class =
+        lean_register_external_class(donothing_finalize, donothing_foreach);
+  }
+  return lean_alloc_external(g_Value_class, s);
+}
+
+static inline LLVMValueRef lean_to_Value(b_lean_obj_arg s) {
+  return static_cast<LLVMValueRef>(lean_get_external_data(s));
+}
+
+// == LLVM <-> Lean: BuilderRef ==
+
+static lean_external_class *g_Builder_class = nullptr;
+
+static inline lean_object *Builder_to_lean(LLVMBuilderRef s) {
+  if (g_Builder_class == nullptr) {
+    g_Builder_class =
+        lean_register_external_class(donothing_finalize, donothing_foreach);
+  }
+  return lean_alloc_external(g_Builder_class, s);
+}
+
+static inline LLVMBuilderRef lean_to_Builder(b_lean_obj_arg s) {
+  return static_cast<LLVMBuilderRef>(lean_get_external_data(s));
+}
+
+// == LLVM <-> Lean: BasicBlockRef ==
+
+static lean_external_class *g_BasicBlock_class = nullptr;
+
+static inline lean_object *BasicBlock_to_lean(LLVMBasicBlockRef s) {
+  if (g_BasicBlock_class == nullptr) {
+    g_BasicBlock_class =
+        lean_register_external_class(donothing_finalize, donothing_foreach);
+  }
+  return lean_alloc_external(g_BasicBlock_class, s);
+}
+
+static inline LLVMBasicBlockRef lean_to_BasicBlock(b_lean_obj_arg s) {
+  return static_cast<LLVMBasicBlockRef>(lean_get_external_data(s));
+}
+
+// == Array Type <-> C array of types ==
+
+// TODO: there is redundancy here, but I am loath to clean it up with templates,
+// because we will somehow dispatch to the correct `lean_to_*` function. I can't
+// think of an immediate, clean way to achieve this (plenty of unclean ways. eg:
+// create a templated function that we partial template specialize that tells us
+// what the correct `lean_to_*` is.
+// Concretely: leanToFn<LLVMTypeRef> = lean_to_LLVMType; leanToFn<LLVMValueRef>
+// = lean_to_LLVMValue).
```

Comment:
Maybe just leave the redundancy? If it is obvious repetitive in a rarely changing file, maybe that's actually a clean approach.

Reply:
Very well. I've removed the `TODO`, but left the comment in-place to point out the redundancy.